### ISSUE TITLE
[prefer-const] False positives when using non-nullish assertion operator

### ISSUE THREAD

**Issue by almeidx** [typescript-eslint/typescript-eslint#2804]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```ts
// ESLint says: 'num' is never reassigned. Use 'const' instead.
let num = data?.a
num! = 1
// same thing happens with +=, -=, etc
```

```JSON
{
  "rules": {
    "prefer-const": "error"
  }
}
```

```TS
// your repro code case
let xp = data?.xp;
/* bunch of unrelated code ... */
xp! = 123;
```

**Expected Result**

Nothing, since `num` is being reassigned

**Actual Result**

ESLint says `'xp' is never reassigned. Use 'const' instead.`

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.6.0` |
| `@typescript-eslint/parser`        | `4.6.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `14.14.0` |

**Comment by bradzacher** [comment]
Your example shows that `num` is never reassigned.
Which is aligned exactly with the error message you've included.

I think maybe you're misunderstanding the lint rule and how `const` works?

The rule isn't saying "num must be assigned to a different variable".
The rule is saying "a different value is never assigned to num".

**Comment by almeidx** [comment]
the error I said in the actual result was the error in the repro code. the error on my use case is `'xp' is never reassigned. Use 'const' instead.`

**Comment by bradzacher** [comment]
Why are you assigning to `xp!` instead of just `xp`?

**Comment by almeidx** [comment]
> Why are you assigning to `xp!` instead of just `xp`?

`xp` is type `number | undefined` because of the optional chain.
If I don't use the non-nullish assertion operator this error appears:
![](https://i.imgur.com/pzXHeKx.png)

**Comment by bradzacher** [comment]
okay... so your example again is incorrect.

Your example should actually be
```ts
let xp: number | undefined = 1;
xp! += 1;
```

If you ensure your initial example is a completely isolated reproduction of your code, then we can save all this back-and-forth.

**Comment by almeidx** [comment]
Alright, sorry for the trouble. 😅
